tests/trainer/test_tp.py

```diff
@@ -1,27 +1,360 @@
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.utils.data import DataLoader, Dataset
 
+from composer.callbacks import MemoryMonitor
+from composer.loggers import InMemoryLogger
 from composer.trainer.trainer import Trainer
-from composer.utils import dist
+from composer.utils import FSDPConfig, ParallelismConfig, TPConfig, dist, reproducibility
 from tests.common import (
     RandomClassificationDataset,
+    RandomClassificationDatasetReplicated,
+    SimpleComposerMLP,
     SimpleModel,
+    deep_compare,
     world_size,
 )
 
 
+def get_trainer(
+    parallelism_config: Optional[ParallelismConfig] = None,
+    size: int = 4,
+    batch_size: int = 1,
+    num_classes: int = 2,
+    num_features: int = 2,
+    seed: int = 44,
+    device: Union[torch.device, str] = 'cuda',
+    replication: Optional[int] = None,
+):
+    """Trainer for a simple model with any parallelism_config."""
+
+    reproducibility.seed_all(seed)
+    if isinstance(device, str):
+        device = torch.device(device)
+
+    dataset: Dataset = RandomClassificationDatasetReplicated(
+        shape=(num_features,),
+        num_classes=num_classes,
+        size=size,
+        device=device,
+        replication=replication,
+    )  # X=(num_features,), y=(,), i.e. scalar
+
+    dataloader = DataLoader(
+        dataset,
+        sampler=dist.get_sampler(dataset),
+        batch_size=batch_size,
+    )  # X=(batch_size, num_features), y=(batch_size,)
+
+    model = SimpleComposerMLP(num_features=num_features, device=device, num_classes=num_classes)
+
+    trainer = Trainer(
+        seed=seed,
+        device='gpu',
+        model=model,
+        max_duration='1ep',
+        train_dataloader=dataloader,
+        precision='fp32',
+        parallelism_config=parallelism_config,
+        callbacks=[MemoryMonitor()],
+        loggers=[InMemoryLogger()],
+        progress_bar=False,
+        log_to_console=False,
+    )
+
+    return trainer
+
+
+def get_ddp_trainer(
+    size: int = 4,
+    batch_size: int = 1,
+    num_classes: int = 2,
+    num_features: int = 2,
+    seed: int = 44,
+    device: Union[torch.device, str] = 'cuda',
+    replication: Optional[int] = None,
+):
+    ddp_trainer = get_trainer(
+        size=size,
+        batch_size=batch_size,
+        num_classes=num_classes,
+        num_features=num_features,
+        seed=seed,
+        device=device,
+        replication=replication,
+    )
+    return ddp_trainer
+
+
+def get_fsdp_trainer(
+    size: int = 4,
+    batch_size: int = 1,
+    num_classes: int = 2,
+    num_features: int = 2,
+    seed: int = 44,
+    device: Union[torch.device, str] = 'cuda',
+    replication: Optional[int] = None,
+):
+    fsdp_config = FSDPConfig(
+        state_dict_type='full',
+        sharding_strategy='SHARD_GRAD_OP',
+        mixed_precision='full',
+        use_orig_params=True,
+    )
+    parallelism_config = ParallelismConfig(fsdp=fsdp_config)
+
```

Comment:
can fsdp + ddp + tp trainers all be one fn that has an arg with parallelism strat to reduce code copying?

Reply:
done!